streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java

```diff
@@ -0,0 +1,838 @@
+                db.newIterator(withTimestampColumnFamily),
+                openIterators,
+                from,
+                to);
+        }
+
+        @Override
+        public KeyValueIterator<Bytes, byte[]> all() {
+            final RocksIterator innerIterWithTimestamp = db.newIterator(withTimestampColumnFamily);
+            innerIterWithTimestamp.seekToFirst();
+            return new RocksDbIterator(name, innerIterWithTimestamp, openIterators);
+        }
+
+        @Override
+        public long approximateNumEntries() throws RocksDBException {
+            return db.getLongProperty(withTimestampColumnFamily, "rocksdb.estimate-num-keys");
+        }
+
+        @Override
+        public void flush() throws RocksDBException {
+            db.flush(fOptions, withTimestampColumnFamily);
+        }
+
+        @Override
+        public void restoreAllInternal(final Collection<KeyValue<byte[], byte[]>> records,
+                                       final WriteBatch batch) throws RocksDBException {
+            for (final KeyValue<byte[], byte[]> record : records) {
+                if (record.value == null) {
+                    batch.delete(withTimestampColumnFamily, record.key);
+                } else {
+                    batch.put(withTimestampColumnFamily, record.key, record.value);
+                }
+            }
+        }
+
+        @Override
+        public void close() {
+            withTimestampColumnFamily.close();
+        }
+    }
+
+    private class DualColumnFamilyAccessor implements RocksDBAccessor {
+
+        @Override
+        public void put(final byte[] rawKey,
+                        final byte[] rawValueWithTimestamp) {
+            if (rawValueWithTimestamp == null) {
+                try {
+                    db.delete(noTimestampColumnFamily, wOptions, rawKey);
+                    db.delete(withTimestampColumnFamily, wOptions, rawKey);
+                } catch (final RocksDBException e) {
+                    // String format is happening in wrapping stores. So formatted message is thrown from wrapping stores.
+                    throw new ProcessorStateException("Error while removing key %s from store " + name, e);
+                }
+            } else {
+                try {
+                    db.delete(noTimestampColumnFamily, wOptions, rawKey);
+                    db.put(withTimestampColumnFamily, wOptions, rawKey, rawValueWithTimestamp);
+                } catch (final RocksDBException e) {
+                    // String format is happening in wrapping stores. So formatted message is thrown from wrapping stores.
+                    throw new ProcessorStateException("Error while putting key %s value %s into store " + name, e);
+                }
+            }
+        }
+
+        @Override
+        public void putAll(final List<KeyValue<Bytes, byte[]>> entries, final WriteBatch batch) throws RocksDBException {
+            for (final KeyValue<Bytes, byte[]> entry : entries) {
```

Comment:
`write(batch)` should be done by the caller -- that we call it in the other accessor is a bug -- will rename the method to `prepareBatch()`.

Reply:
Not sure I can follow this: could you elaborate a bit more? I was thinking to use the RocksDB's

```
void write(final WriteOptions writeOpts, final WriteBatch updates)
```

API for the putAll call, is it possible?

Reply:
That is exactly what is use. Compare `RocksDB#putAll()`:
```
    @Override
    public void putAll(final List<KeyValue<Bytes, byte[]>> entries) {
        try (final WriteBatch batch = new WriteBatch()) {
            dbAccessor.prepareBatch(entries, batch);
            write(batch);
        } catch (final RocksDBException e) {
            throw new ProcessorStateException("Error while batch writing to store " + name, e);
        }
    }
```

It calls `dbAccessor.prepareBatch(entries, batch);` (already renamed from `putAll -> prepareBatch`) that is this method. We do the call "outside" because only who the batch is prepared differs.

Does this make sense